source/data-modeling/field-types.txt

```diff
@@ -0,0 +1,558 @@
+  instance of your custom type.
+- ``evolve``, which takes an instance of your custom type and converts is to a
+  criteria that MongoDB can use to query the database.
+
+The following example creates a custom field type called ``Point`` and
+implements the preceding methods:
+
+.. literalinclude:: /includes/data-modeling/field-types.rb
+   :language: ruby
+   :start-after: # start-custom-field-type
+   :end-before: # end-custom-field-type
+
+In the preceding example, the ``mongoize`` *instance method* accepts an instance
+of your custom type object and converts it to an ``Array`` to store in the
+database. The ``mongoize`` *class method* accepts objects of all types and
+converts them to similar types that can be stored in the database. {+odm+} uses
+the ``mongoize`` class method when it calls the getter and setter methods.
+
+The ``demongoize`` method converts the stored ``Array`` value into the custom
+``Point`` type. The {+odm+} uses this method when it calls the getter.
+
+The ``evolve`` method converts the custom ``Point`` type into a queryable
+``Array`` type, and converts all other types to ``object``. {+odm+} uses this
+method when it calls a method that queries the database.
+
+Phantom Custom Field Types
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can create a custom field type that saves a different value to the database
+than the value assigned in the application. This can be useful to have
+descriptive values in the application while storing more compact values in the
+database.
+
+The following example creates a ``ColorMapping`` type that uses the name of the
+color in the application, but stores the color as an integer in the database:
+
+.. literalinclude:: /includes/data-modeling/field-types.rb
+   :language: ruby
+   :start-after: # start-phantom-field-type
+   :end-before: # end-phantom-field-type
+
+Dynamic Fields
+--------------
+
+You can instruct {+odm+} to create fields dynamically by setting the
+``Mongoid::Attributes::Dynamic`` module in your model. This allows {+odm+} to
+create fields based on an arbitrary hash, or by the documents stored in the
+database.
+
+The following example creates a ``Person`` class with dynamic fields:
+
+.. literalinclude:: /includes/data-modeling/field-types.rb
+   :language: ruby
+   :start-after: # start-dynamic-field
+   :end-before: # end-dynamic-field
+
+You can also specify both fixed fields and dynamic fields within the same class.
+When doing so, {+odm+} treats all attributes for properties with field definitions
+according to their field type, and all other attributes as dynamic.
+
+When using dynamic fields in your application, you must initially set the value in one of the
+following ways:
+
+- Pass the attribute hash to the constructor.
+- Assign values by using the ``attributes=`` method.
+- Assign values by using the ``[]=`` method.
+- Assign values by using the ``write_attribute`` method.
+- Work with values that are already present in the database.
+
```

Comment:
S: provide example?

Reply:
Since these are all just standard ways to assign values in mongoid, I didn't think an example would add much value and opted to keep it off since this page is long already